pipelines.md

```diff
@@ -0,0 +1,358 @@
+sketch007 = sketch006
+  |> lineTo([profileStartX(sketch006), profileStartY(sketch006)])
+  |> close()
+  |> revolve({ axis: 'y' })
+```
+
+See below ('Tagging') for a further improvement to this example.
+
+#### Alternative
+
+Retain `#` or replace it with a keyword (which I would prefer) for the rare cases where it is required (c.f., the current situation where it is frequently required).
+
+## Tagging
+
+Remove the `$` operator, remove the concept of passing tags to functions. Allow the `as` keyword to be used within pipelines to assign intermediate results to a variable. Note that this usage of `as` follows its use in `import` statements and has the same semantics of introducing a new name.
+
+This syntax is easier to read and comprehend, extends to user-defined functions, and reduces the total feature count of the language (since it is already used in imports).
+
+Example: `foo = ... |> bar() as baz |> qux()`, here the final result is assigned into `foo` (no change to semantics), the result of executing `|> bar()` is assigned into `baz` (equivalent to `|> bar($baz)` in the current syntax).
+
+The earlier example becomes:
+
+```
+sketch006 = startSketchOn('XZ')
+  |> startProfileAt([0.1, 1])
+  |> line([0.1, 0])
+  |> angledLineToX({ angle: 10, to: 0.05 })
+  |> yLine(10)
+  |> line([0.6, 0])
+  |> yLine(-.05)
+  |> tangentialArc({ radius: 0.6, offset: -90 }) as arcSketch
+  |> lineTo([profileStartX(arcSketch), profileStartY(arcSketch)])
+  |> close()
+  |> revolve({ axis: 'y' })
+```
+
+### Alternative
+
+Rather than making `as` part of the pipeline syntax, it could be allowed in any expression. The above description would still apply, but it could also be applied to sub-expressions, e.g., `hole(circle(...) as innerCircle)`.
+
+This would be more flexible and would avoid learners hitting the issue of knowing where `as` can be used ("I can use it here and here, so why can't I use it here?"). However, I believe it would lead to poor programming style: it is generally better to use a variable declaration to refer to reused geometry or data rather than use `as` since it will be easier to scan code for declarations and it encourages a less nested, more straightforward coding style. On the other hand, pipelining is ergonomic and allowing `as` there prevents breaking pipelines arbitrarily.
+
+The precedence in general sub-expressions may be confusing, e.g., in `a + b as c + d`, does `c` refer to the value of `b` or of `a + b`? 
+
+## Braced blocks in pipelines
+
+Braced blocks are allowed in pipelines. The current receiver of the pipeline is piped into the *last* expression of the block (the block must have a last expression, similar to `if` blocks). The result of the block is the result of the final expression. Variables within the block are scoped to the block. (Possible extension: use `export x = ...` or `... export as x` to make variables visible in the enclosing scope).
+
+```
+startSketchOn('XZ')
+  |> {
```

Comment:
Yes, but with the addition of having the subject of the pipeline passed to the last expression of the block.

I'd prefer not to reuse the anon fn syntax here, it would be really noisy. I would like to think a bit about what functions *are* and how the user thinks of them and how much we rely on them (Jon has rethought this too in his constraints proposal). I'm kind of hoping we can get rid of anon functions entirely since they are a pretty hard to grok concept for non-programmers, but that's a bit of a tangent...